src/tools/weave/Cmd_PrintAccessToken.cpp

```diff
@@ -0,0 +1,193 @@
+ *      that decodes and prints the contents of a Weave access token certificate.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <getopt.h>
+#include <errno.h>
+#include <unistd.h>
+
+#include "weave-tool.h"
+#include <Weave/Support/ASN1OID.h>
+#include <Weave/Profiles/security/WeaveSecurityDebug.h>
+#include <Weave/Profiles/security/WeaveAccessToken.h>
+#include <Weave/Support/Base64.h>
+
+using namespace nl::Weave::ASN1;
+using namespace nl::Weave::Profiles::Security;
+
+#define CMD_NAME "weave print-cert"
+
+static bool HandleNonOptionArgs(const char *progName, int argc, char *argv[]);
+static bool HandleOption(const char *progName, OptionSet *optSet, int id, const char *name, const char *arg);
+
+static OptionDef gCmdOptionDefs[] =
+{
+    { "base64", kNoArgument, 'b' },
+    { }
+};
+
+static const char *const gCmdOptionHelp =
+    "   -b, --base64\n"
+    "\n"
+    "       The file containing the TLV should be parsed as base64.\n"
+    "\n"
+    ;
+
+static OptionSet gCmdOptions =
+{
+    HandleOption,
+    gCmdOptionDefs,
+    "COMMAND OPTIONS",
+    gCmdOptionHelp
+};
+static HelpOptions gHelpOptions(
+    CMD_NAME,
+    "Usage: " CMD_NAME " [<options...>] <cert-file>\n",
+    WEAVE_VERSION_STRING "\n" COPYRIGHT_STRING,
+    "Print a Weave Access Token certificate.\n"
+    "\n"
+    "ARGUMENTS\n"
+    "\n"
+    "  <access-token>\n"
+    "\n"
+    "       A file containing a Weave Access Token, in base-64 format\n"
+    "\n"
+);
+
+static OptionSet *gCmdOptionSets[] =
+{
+    &gCmdOptions,
+    &gHelpOptions,
+    NULL
+};
+
+static const char *gCertFileName = NULL;
+static bool gUseBase64Decoding = false;
+enum {
+    kNumCerts = 3,
+    kCertBufSize = 1024,
+};
+
+bool Cmd_PrintAccessToken(int argc, char *argv[])
+{
+    bool res = true;
+    WEAVE_ERROR err;
+    WeaveCertificateData *certData;
+    WeaveCertificateSet certSet;
+    uint8_t * accessToken = NULL;
+    uint32_t accessTokenLen = 0;
+    TLVReader reader;
+
+    if (argc == 1)
+    {
+        gHelpOptions.PrintBriefUsage(stderr);
+        ExitNow(res = true);
+    }
+
+    if (!ParseArgs(CMD_NAME, argc, argv, gCmdOptionSets, HandleNonOptionArgs))
+    {
+        ExitNow(res = false);
+    }
+
+    if (!ReadFileIntoMem(gCertFileName, accessToken, accessTokenLen))
+        ExitNow(res = false);
+
+    if (gUseBase64Decoding)
+    {
+        uint32_t b64len = accessTokenLen;
+        uint8_t * b64 = accessToken;
+        accessToken = (uint8_t *)malloc(accessTokenLen);
+        accessTokenLen = nl::Base64Decode((const char *)b64, b64len, accessToken);
+        free(b64);
+    }
+
+    reader.Init(accessToken, accessTokenLen);
+    certSet.Init(kNumCerts, kCertBufSize);
+    err = LoadAccessTokenCerts(reader, certSet, 0, certData);
+
+    if (err != WEAVE_NO_ERROR)
+    {
+        fprintf(stderr, "Error reading cert info: %s.\n", nl::ErrorStr(err));
+        ExitNow(res = false);
+    }
+
+    err = DetermineCertType(*certData);
```

Comment:
Is this necessary? For access tokens, this function sets the type as kCertType_General.